### PR TITLE
Refactor to centralize GUIButton styling/rendering code

### LINKED ISSUE #8986: Button in Minimal fails clipping test
##### Minetest version
```
26b39f1eae1f576669cbf49c6db94ef4ed8624df
```

##### Steps to reproduce
1. In Minimal, say `/formspec`
2. Select “Noclip” tab
3. Look at the 3rd button from the top, at the left

I expect the text on this button to be clipped.
But in reality, it is visible.

### PR DESCRIPTION
**Pull request by Df458**
(This addresses the first point in #9044)

This PR refactors some GUIButton-related code, mainly code related to `image_button` and `item_image_button`. These controls are now subclasses of GUIButton, and handle their custom style properties and content rendering via overloads. To do this, style property application has also been moved into GUIButton.

## Benefits

1. By centralizing rendering and styling code, we can be reasonably certain that future additions and changes to styling and rendering will be applied to alternate button types with minimal development time
2. By moving `item_image_button`'s rendering code into the control itself, I inadvertently fixed #8986 (At least, I fixed the buttons. The other labels that were pointed out in the comments are not fixed). Personally, I think this bug would have been unlikely to occur in the first place had item_image_button not been designed as multiple controls layered on top of each other and I expect to see further issues of this sort if the design remains.
3. This change makes **all** style properties that apply to `button` automatically apply to `image_button` and `item_image_button`. It also ensures that future additions will apply without extra consideration or code duplication.
4. This change gives `image_button` a distinct purpose. `button`'s `bgimg` property had rendered `image_button` mostly obsolete, this change creates a distinction between foreground and background images and allows `image_button` to have both. The result is that custom image buttons no longer need to have a background baked into their texture; it can be supplied via a formspec prepend. Just for good measure, I also added a full set of style properties for `image_button`'s foreground image, allowing complete visual feedback here as well.

## Impact
From my testing this should have no appreciable impact on existing mods, my goal was to keep the default behavior as close as possible to what we currently have. Since all new features are handled via style properties, mods that take advantage of the new features should still be compatible with older versions of Minetest.

If you find any notable differences while testing, please report them!

## Testing
The easiest way to test the impact of this PR is to compare various formspecs in the current release against a patched copy. To showcase some of the new possibilities, I also made a small demo mod. The demo formspec opens when you join a world, or when using the `/buttons` command. You can get it [here](https://github.com/minetest/minetest/files/3817217/button_test.zip).

**UPDATE**: In anticipation of approval/merging, I've added a set of buttons demonstrating the new style properties to the test formspec as well. These will also work for testing, and you can use the `/formspec` command to access them.

### CODE REVIEW

**Comment by Df458** [comment]
I'm not sure I agree with some of the suggestions coming out of clag-format... I'd like the opinion of an approver. Should I follow all of them? If not, are there any remaining style changes I should make?

**Comment by sfan5** [comment]
Most remaining suggestions don't look very useful, you can add guiButtonImage.cpp and guiButtonItem.cpp to the whitelist at `util/travis/clang-format-whitelist.txt`.

**Comment by Df458** [comment]
I've rebased on top of #8740, but you may notice that I left part of one conflict commented. This won't prevent it from working as intended, it's just a reminder to myself since I'd like to properly involve the new GUIItemImage class with GUIButtonItem. I didn't see an option to mark the PR as WIP again, so for now just bear in mind that this part of the code may change in the next couple of days.

**Comment by Df458** [comment]
Alright, despite some irl issues I've finally managed to sit down and finish updating this PR. The main change is explained below.

In order to better work with the new `GUIItemImage` class, I decided to make `GUIButton` (and derived classes) act as a proper container for child elements, rather than offering specialized rendering for derived classes, the new classes instead handle any additional style properties and update their children in response to property changes.

The result looks mostly identical to what we have before, but there are some notable changes in the back:

- `GUIItemImage` no longer needs to know or care what its parent is, because `GUIButton` automatically offsets all child elements as necessary
- `GUIButton`'s label is now a child `StaticText` element, and follows the same offsets as other content (as opposed to the separate offsets in the skin). **I consider this to be a feature**, not a bug, because it makes the text and content move together in a consistent manner.
- To make this work, I had to modify `StaticText` to get around an issue with the text color when rendering. I'm not sure if that change was correct, I welcome anyone to help test this (and offer alternatives if this breaks something else that I missed on my end).
- `GUIButtonItem` isn't totally necessary anymore... I decided to keep it regardless, because there's at least one near-future PR on my todo list that requires it and I think it makes more logical sense to add it during this refactor. I am open to changing this if the core devs disagree.

Those are the main diversions from the original PR. The result is the same, but they also open up a couple of new avenues for future work:

- We could add a button formspec element with mod-defined content now, and have it properly handle its children.
-  We could (and should if this is merged) add style properties for defining custom offsets for content, ideally with hovered/pressed variants. One thing that you can see in my test formspec is that most of the big custom buttons don't have their child elements moving as much as the button's surface implies. This is an existing limitation, but we'll be in a position to easily correct it once this is merged.

Like most cool possibilities, I've elected not to implement these here for the sake of keeping the PR focused. Since I'm on this strange months-long button crusade, I'll probably wind up making PRs for some of these features in the near future.

**Comment by Df458** [comment]
Sorry to bother y'all, but this has been sitting ready for about a week. Any chance of a review sometime soon?

**Comment by Enrikoo** [comment]
To someone who doesn't want to compile this, here is an extra package for win64:
[minetest-5.2.0-4640c988-win64.zip](https://github.com/minetest/minetest/files/3912017/minetest-5.2.0-4640c988-win64.zip)

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2535,13 +2535,10 @@ Some types may inherit styles from parent types.
...
     * noclip - boolean, set to true to allow the element to exceed formspec bounds.
-* image_button
-    * alpha - boolean, whether to draw alpha in bgimg. Default true.
-    * border - boolean, draw border. Set to false to hide the bevelled button pane. Default false.
-    * noclip - boolean, set to true to allow the element to exceed formspec bounds.
-* item_image_button
-    * border - boolean, draw border. Set to false to hide the bevelled button pane. Default false.
-    * noclip - boolean, set to true to allow the element to exceed formspec bounds.
+* image_button (additional properties)
+    * fgimg - standard image. Defaults to none.
```
Comment (rubenwardy) [inline doc/lua_api.txt]: State which image take precedence (it should probably be the element one)

File: src/gui/StyleSpec.h [hunk 2]
```diff
@@ -106,6 +117,29 @@ class StyleSpec
...
 
+	video::ITexture *getTexture(Property prop, ISimpleTextureSource *tsrc,
+			video::ITexture *def) const
+	{
+		const auto &val = properties[prop];
+		if (val.empty()) {
+			return def;
+		}
+
+		video::ITexture *texture = tsrc->getTexture(val);
```
Comment (rubenwardy) [inline src/gui/StyleSpec.h]: Shouldn't this use a texture source? `ITextureSource`, I think it's called

Test with texture modifiers

File: src/gui/guiButtonImage.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	virtual void setScaleImage(bool scaleImage=true) override;
+
+	//! Do not drop returned handle
+	static GUIButtonImage *addButton(gui::IGUIEnvironment *environment,
+			const core::rect<s32> &rectangle, IGUIElement *parent, s32 id,
+			const wchar_t *text, const wchar_t *tooltiptext = L"");
+
+private:
+	ButtonImage ButtonForegroundImages[gui::EGBIS_COUNT];
```
Comment (rubenwardy) [inline src/gui/guiButtonImage.h]: names starting with lowercase would be preferred

File: src/gui/guiButtonItem.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#include "client/client.h"
+#include "client/hud.h" // drawItemStack
+#include "guiItemImage.h"
+#include "IGUIEnvironment.h"
+#include "itemdef.h"
+
+using namespace irr;
+using namespace gui;
+
+GUIButtonItem::GUIButtonItem(gui::IGUIEnvironment *environment, gui::IGUIElement *parent,
```
Comment (rubenwardy) [inline src/gui/guiButtonItem.cpp]: GUIButtonItemImage

**Comment by rubenwardy** [comment]
Note: updating the formspec test could just be modifying your demo mod to instead add a tab to the `minimal/test/formspec.lua` formspec

File: doc/lua_api.txt (hunk 1 above)
Comment (Df458) [reply doc/lua_api.txt]: Good catch actually, the element-supplied textures were not taking precedence but should've been. Other style properties were already subject to this, but I missed those two.

**Comment by Df458** [comment]
I noticed that there was a free row on the styles page, so I've added a pair of fully kitted-out buttons there. I can still change this to another page if you'd prefer, but I figure this is good enough as a minimal example.

**Comment by rubenwardy** [comment]
> I can still change this to another page if you'd prefer, but I figure this is good enough as a minimal example.

The idea is that the page demonstrates a wide range of different elements, so that it can be checked that styles work and things are rendered as expected. It doesn't need to be exhaustive but it should be a wide enough range. I haven't check what you've added yet, I'm quite tired and would like to do that another day

Because you added `.cpp` files, you'll need to update Android.mk. (Hopefully that file will go away soon, when CMake is used to build on Android)

https://github.com/minetest/minetest/blob/master/build/android/jni/Android.mk#L183

**Comment by rubenwardy** [comment]
Oops, I meant to clear the "changes requested" flag on Github

**Comment by Df458** [comment]
Take your time :)
It's better to take a little longer to ensure that the change is good than to rush.

**Pull request merged**